File: erlang/mptcp.erl
%%% @author Anthony Doerane <anthony.doeraene@hotmail.com>

-module(mptcp).
-export([client/2, server/2]).

-define(IPPROTO_MPTCP, 262).

open_socket() ->
   {Status, Sock} = socket:open(inet, stream, ?IPPROTO_MPTCP),
   case Status of
      ok -> {Status, Sock};
      _ -> socket:open(inet, stream, 0)
   end.

client(SAddr, SPort) ->
   {ok, Sock} = open_socket(),
   ok = socket:connect(Sock, #{family => inet,
                               addr   => SAddr,
                               port   => SPort}),
   Msg = <<"hello">>,
   ok = socket:send(Sock, Msg),
   ok = socket:shutdown(Sock, write),
   {ok, Msg} = socket:recv(Sock),
   ok = socket:close(Sock).

server(Addr, Port) ->
   {ok, LSock} = open_socket(),
   ok = socket:bind(LSock, #{family => inet,
                             port   => Port,
                             addr   => Addr}),
   ok = socket:listen(LSock),
   {ok, Sock} = socket:accept(LSock),
   {ok, Msg} = socket:recv(Sock),
   ok = socket:send(Sock, Msg),
   ok = socket:close(Sock),
   ok = socket:close(LSock).

File: elixir/lib/mptcp.ex
# Author: Anthony Doeraene <anthony.doeraene@hotmail.com>

defmodule MPTCP do
  @ipproto_mptcp 262

  defp open_socket() do
    {status, sock} = :socket.open(:inet, :stream, @ipproto_mptcp)
    case status do
      :ok -> {:ok, sock}
      _ -> :socket.open(:inet, :stream, 0)
    end
  end

  def client(addr, port) do
    {:ok, sock} = open_socket()
    :ok = :socket.connect(sock, %{family: :inet,
                                addr: addr,
                                port: port})
    msg = <<"hello">>
    :ok = :socket.send(sock, msg)
    :ok = :socket.shutdown(sock, :write)
    {:ok, msg} = :socket.recv(sock)
    :ok = :socket.close(sock)
    msg
  end

  def server(addr, port) do
    {:ok, lsock} = open_socket()
    :ok = :socket.bind(lsock, %{family: :inet,
                              port: port,
                              addr: addr})
    :ok = :socket.listen(lsock)
    {:ok, sock} = :socket.accept(lsock)
    {:ok, msg} = :socket.recv(sock)
    :ok = :socket.send(sock, msg)
    :ok = :socket.close(sock)
    :ok = :socket.close(lsock)
  end
end
